### PR TITLE
Implemented decomposable attention model

### PR DESCRIPTION
**Pull request by matt-gardner**
This depends on #64, #65, and #66, and still needs some tests, but is pretty close to done.

### CODE REVIEW

File: allennlp/models/decomposable_attention.py [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+from allennlp.models.model import Model
+from allennlp.modules import FeedForward, MatrixAttention
+from allennlp.modules import Seq2SeqEncoder, SimilarityFunction, TimeDistributed, TextFieldEmbedder
+
+
+@Model.register("decomposable_attention")
+class DecomposableAttention(Model):
+    """
+    This ``Model`` implements the Decomposable Attention model described in "A Decomposable
+    Attention Model for Natural Language Inference", by Parikh et al., 2016, with some optional
```
Comment (DeNeutoy) [inline allennlp/models/decomposable_attention.py]: Add a link to the paper: https://arxiv.org/abs/1606.01933

File: allennlp/models/decomposable_attention.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+    enhancements before the decomposable attention actually happens.  Parikh's original model
+    allowed for computing an "intra-sentence" attention before doing the decomposable entailment
+    step.  We generalize this to any :class:`Seq2SeqEncaoder` that can be applied to the premise
```
Comment (DeNeutoy) [inline allennlp/models/decomposable_attention.py]: /s/`Encaoder`/`Encoder`

File: allennlp/models/decomposable_attention.py [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+
+        label_logits : torch.FloatTensor
+            A tensor of shape ``(batch_size, num_labels)`` representing unnormalised log
+            probabilities of the entailment label.
+        label_probs : torch.FloatTensor
+            A tensor of shape ``(batch_size, num_labels)`` representing probabilities of the
+            entailment label.
+        loss : torch.FloatTensor, optional
+            A scalar loss to be optimised.
+
```
Comment (DeNeutoy) [inline allennlp/models/decomposable_attention.py]: remove extra space.

File: allennlp/models/decomposable_attention.py [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+        h2p_attention = last_dim_softmax(similarity_matrix.transpose(1, 2).contiguous(), premise_mask)
+        # Shape: (batch_size, hypothesis_length, embedding_dim)
+        attended_premise = weighted_sum(embedded_premise, h2p_attention)
+
+        premise_compare_input = torch.cat([embedded_premise, attended_hypothesis], dim=-1)
+        hypothesis_compare_input = torch.cat([embedded_hypothesis, attended_premise], dim=-1)
+
+        compared_premise = self._compare_feedforward(premise_compare_input)
+        # TODO(mattg): use broadcasting once pytorch 0.2 is released.
+        print("compared_premise:", compared_premise)
```
Comment (DeNeutoy) [inline allennlp/models/decomposable_attention.py]: remove print statements.

File: allennlp/models/decomposable_attention.py [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+        -------
+        A Dict containing:
+
+        label_probs : torch.FloatTensor
+            A tensor of shape ``(num_labels,)`` representing probabilities of the entailment label.
+        """
+        instance = Instance({"premise": premise, "hypothesis": hypothesis})
+        instance.index_fields(self._vocab)
+        model_input = arrays_to_variables(instance.as_array(instance.get_padding_lengths()))
+        for tensor in model_input["premise"].values():
```
Comment (DeNeutoy) [inline allennlp/models/decomposable_attention.py]: Remove and add the `add_batch_dimension` flag to `arrays_to_variables`

File: allennlp/models/decomposable_attention.py [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+    similarity_function : ``SimilarityFunction``
+        This is the similarity function used when computing the similarity matrix between words in
+        the premise and words in the hypothesis.
+    compare_feedforward : ``FeedForward``
+        This feedforward network is applied to the aligned premise and hypothesis representations,
+        individually.
+    aggregate_feedforward : ``FeedForward``
+        This final feedforward network is applied to the concatenated, summed result of the
+        ``compare_feedforward`` network, and its output is used as the entailment class logits.
+    premise_encoder : ``Seq2SeqEncoder``, optional (default=``None``)
```
Comment (DeNeutoy) [inline allennlp/models/decomposable_attention.py]: What is the motivation for adding this in? Is there some extension paper which added encoders or something? Seems to unnecessarily clutter the implementation and it isn't in the paper.

Comment (matt-gardner) [inline allennlp/models/decomposable_attention.py]: The paper had an intra-sentence attention mechanism that improved performance, and could be implemented as a seq2seq encoder.  Other follow-on work has also used a bi-LSTM to encode the sentences before doing a very similar computation (talk to Yang, for example).  Implementing the model this way lets us easily handle either of those extensions, without much additional complexity.

Comment (DeNeutoy) [inline allennlp/models/decomposable_attention.py]: Cool.

**Pull request merged**